src/cryptography/hazmat/bindings/openssl/binding.py

```diff
@@ -4,136 +4,30 @@
 
 from __future__ import absolute_import, division, print_function
 
-import os
-import sys
 import threading
 
-from cryptography.hazmat.bindings.utils import (
-    build_ffi_for_binding, load_library_for_binding,
-)
-
-
-_OSX_PRE_INCLUDE = """
-#ifdef __APPLE__
-#include <AvailabilityMacros.h>
-#define __ORIG_DEPRECATED_IN_MAC_OS_X_VERSION_10_7_AND_LATER \
-    DEPRECATED_IN_MAC_OS_X_VERSION_10_7_AND_LATER
-#undef DEPRECATED_IN_MAC_OS_X_VERSION_10_7_AND_LATER
-#define DEPRECATED_IN_MAC_OS_X_VERSION_10_7_AND_LATER
-#endif
-"""
-
-_OSX_POST_INCLUDE = """
-#ifdef __APPLE__
-#undef DEPRECATED_IN_MAC_OS_X_VERSION_10_7_AND_LATER
-#define DEPRECATED_IN_MAC_OS_X_VERSION_10_7_AND_LATER \
-    __ORIG_DEPRECATED_IN_MAC_OS_X_VERSION_10_7_AND_LATER
-#endif
-"""
-
-
-def _get_libraries(platform):
-    # OpenSSL goes by a different library name on different operating systems.
-    if platform != "win32":
-        # In some circumstances, the order in which these libs are
-        # specified on the linker command-line is significant;
-        # libssl must come before libcrypto
-        # (http://marc.info/?l=openssl-users&m=135361825921871)
-        return ["ssl", "crypto"]
-    else:
-        link_type = os.environ.get("PYCA_WINDOWS_LINK_TYPE", "static")
-        return _get_windows_libraries(link_type)
-
-
-def _get_windows_libraries(link_type):
-    if link_type == "dynamic":
-        return ["libeay32", "ssleay32", "advapi32"]
-    elif link_type == "static" or link_type == "":
-        return ["libeay32mt", "ssleay32mt", "advapi32",
-                "crypt32", "gdi32", "user32", "ws2_32"]
-    else:
-        raise ValueError(
-            "PYCA_WINDOWS_LINK_TYPE must be 'static' or 'dynamic'"
-        )
+from cryptography._cryptography_openssl import ffi, lib
+
+res = lib.Cryptography_add_osrandom_engine()
```

Comment:
I think I prefer locks. Doing things at module import time is pretty gross
to me.

On Sun, Jun 7, 2015 at 10:00 AM, Paul Kehrer <notifications@github.com>
wrote:

> In src/cryptography/hazmat/bindings/openssl/binding.py
> <https://github.com/pyca/cryptography/pull/1986#discussion_r31877036>:
>
> > -        return _get_windows_libraries(link_type)
> > -
> > -
> > -def _get_windows_libraries(link_type):
> > -    if link_type == "dynamic":
> > -        return ["libeay32", "ssleay32", "advapi32"]
> > -    elif link_type == "static" or link_type == "":
> > -        return ["libeay32mt", "ssleay32mt", "advapi32",
> > -                "crypt32", "gdi32", "user32", "ws2_32"]
> > -    else:
> > -        raise ValueError(
> > -            "PYCA_WINDOWS_LINK_TYPE must be 'static' or 'dynamic'"
> > -        )
> > +from cryptography._cryptography_openssl import ffi, lib
> > +
> > +res = lib.Cryptography_add_osrandom_engine()
>
> OpenSSL lock initialization and osrandom engine addition need to happen
> once and only once. Previously we accomplished this with locks, but it
> seems much cleaner to do it directly on import. Is there another approach
> you'd prefer?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/1986/files#r31877036>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084